File: 프로그래머스/3/43162. 네트워크/네트워크.cpp
#include <bits/stdc++.h>

using namespace std;


vector<int> parents;

int Find(int x){
    if (x == parents[x])
        return x;
    
    return parents[x] = Find(parents[x]);
}

void Union(int a, int b){
    a = Find(a);
    b = Find(b);
    
    if (a < b)
    {
        parents[b] = a;
    } else {
        parents[a] = b;
    }
}



int solution(int n, vector<vector<int>> computers) {
    int answer = 0;
    
    parents = vector<int> (n);
    
    for(int i = 0; i < n; ++i){
        parents[i] = i;
    }
    

    for(int i = 0; i < computers.size(); ++i) {
        for(int j = 0; j < computers[i].size(); ++j){
            if (computers[i][j] == 1)
                Union(i, j);
        }
    }
    
    set<int> s;
    
    for(int i = 0; i < n; ++i) {
        s.insert(Find(i));
    }
    
    answer = s.size();
    
    return answer;
}

File: 프로그래머스/3/43162. 네트워크/네트워크.java

import java.util.*;
import java.util.stream.Collectors;

public class Solution {

    public static int[] parents;

    public static int solution(int n, int[][] computers) {
        int answer = 0;
        parents = new int[n];

        for (int i = 0; i < n; i++) {
            parents[i] = i;
        }
        for (int i = 0; i < computers.length; i++) {
            for (int j = 0; j < computers[i].length; j++) {
                if(computers[i][j] == 1) {
                    if (find(i) != find(j))
                        union(i, j);
                }
            }
        }
        
        for(int i = 0; i < computers.length; ++i)
        {
            find(i);
        }
        
        Set<Integer> collect = Arrays.stream(parents).boxed().collect(Collectors.toSet());
        return collect.size();
    }

    public static void union(int a, int b) {
        a = find(a);
        b = find(b);

        if (a < b) {
            parents[b] = a;
        } else {
            parents[a] = b;
        }
    }

    public static int find(int n) {
        if (n != parents[n]) {
            parents[n] = find(parents[n]);
        }

        return parents[n];
    }
}
